### PR TITLE
Introduce Metrics

### LINKED ISSUE #78: Evaluation/Metrics approach
This has been discussed a bit in other issues, but I wanted to make a dedicated issue for us to discuss this as I think it's very important we get this right.

## Background
Some previous discussions [here](https://github.com/pytorch/ignite/pull/54#issuecomment-363231705) and [here](https://github.com/pytorch/ignite/issues/72).

Throughout, I am going to use the motivating example of training a supervised model where you periodically want to compute some metrics against a validation set.

## Current Setup
Currently, in order to accomplish this, you need to do the following:

1. create an `Evaluator`
2. register the `Evaluate` handler to run the `Evaluator` on the validation set and store the predictions in the history
3. add another event handler to actually use this history to compute the metrics you care about
4. log/plot these metrics however you choose

In code, this looks something like this:

```python
model = ...
validation_loader = ...
trainer = ...
evaluator = create_supervised_evaluator(model, cuda=True)
trainer.add_event_handler(Events.EPOCH_COMPLETED, Evaluate(evaluator, validation_loader, epoch_interval=1))
@trainer.on(Events.EPOCH_COMPLETED)
def log(engine):
    print(engine.current_epoch, categorical_accuracy(evaluator.history))
```

### Pros
- keeps library code cleanly separated with minimal implicit dependencies
- user doesn't have to write much code

### Cons
- can be confusing what happens where
  - we have both an `Evaluator` and an `Evaluate` and yet neither one computes any sort of metrics
- there are a lot of ways this could go wrong
  - you have to understand the contract between what gets stored in the `Evaluator`'s history and the metrics functions
  - you have to make sure you attach the `Evaluate` handler and any logging handlers to the same event

## Goals
Evaluating a model is gonna be something that (essentially) everyone does so I think we need to have a good story here. Imo, we should make the supervised case be super easy while still making it possible for non-supervised cases. That being said, I think we want to accomplish this without removing flexibility and without adding a ton of code.

## Ideas
Working backward from what I would like the api to be, it might be nice if you could just do something like this:

```python
model = ...
validation_loader = ...
trainer = ...
@trainer.on(Events.EPOCH_COMPLETED)
def evaluate(engine):
  results = evaluate(model, {'acc': categorical_accuracy})
  # do something with those results
```

It'd be even nicer if I could do something like this:
```python
model = ...
validation_loader = ...
trainer = ...
trainer.add_event_handler(Events.EPOCH_COMPLETED, Evaluate(model, {'acc': categorical_accuracy}))
```

But without making assumptions about how users want to plot/log their evaluation results, this isn't possible. 

What do you all think? Anything here you take issue with? Any ideas on how we can best accomplish this? Do we need to make plotting/logging part of this discussion as well?

### PR DESCRIPTION
**Pull request by jasonkriss**
Resolves #78 

Depends on #83.

This allows us to do the following:

```python
model = ...
training_loader = ...
validation_loader = ...
trainer = ...
evaluator = create_supervised_evaluator(model, metrics={'acc': CategoricalAccuracy()})
@trainer.on(Events.EPOCH_COMPLETED)
def eval(state):
  training_acc = evaluator.run(training_loader).metrics['acc']
  validation_acc = evaluator.run(validation_loader).metrics['acc']
  print('epoch:', state.epoch, 'training acc:', training_acc, 'validation acc:', validation_acc)
```

This PR currently includes `MSE` and `CategoricalAccuracy` metrics. Once we agree everything looks good, I will add the rest of the metrics.

### CODE REVIEW

**Comment by alykhantejani** [comment]
@jasonkriss sorry for the delay on this review - will try to get to it today

File: ignite/metrics/metric.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    def started(self, engine, state):
+        self.reset()
+
+    def iteration_completed(self, engine, state):
+        self.update(state.output)
+
+    def completed(self, engine, state):
+        state.metrics.append(self.compute())
+
+    def attach(self, engine):
```
Comment (alykhantejani) [inline ignite/metrics/metric.py]: what about having an interface similar to the timer here ([see here](https://github.com/pytorch/ignite/blob/master/ignite/handlers/timing.py#L84)) i.e.:

```python
def attach(start=Events.STARTED, update=Events.ITERATION_COMPLETED, compute=Events.COMPLETED, reset=None)
```

Comment (alykhantejani) [reply ignite/metrics/metric.py]: this way we wouldn't need `started()` and `reset()` or `iteration_completed()` and `update()`,

File: ignite/metrics/metric.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        pass
+
+    def started(self, engine, state):
+        self.reset()
+
+    def iteration_completed(self, engine, state):
+        self.update(state.output)
+
+    def completed(self, engine, state):
+        state.metrics.append(self.compute())
```
Comment (alykhantejani) [inline ignite/metrics/metric.py]: not sure this is the best approach? What if we have several metrics, `state.metrics` should probably be a dict which is accessed with the name of the metric? which can be passed in to the constructor (and have a default value)?

Comment (jasonkriss) [reply ignite/metrics/metric.py]: Yea I wasn't sure on this one. I'm def open to the dict approach. The only reason I went with the list is for the simpler API, both in the constructor and for accessing the result. I see two options if we want to use a more explicit approach:

Option 1
```python
evaluator = create_supervised_evaluator(model, metrics={'acc': CategoricalAccuracy()})
```

Option 2
```python
evaluator = create_supervised_evaluator(model, metrics=[CategoricalAccuracy(name='acc')])
```

What do you think?

File: ignite/metrics/metric.py (hunk 1 above)
Comment (jasonkriss) [reply ignite/metrics/metric.py]: I'm not sure if the flexibility of passing in the events is something we need (or even want) for metrics. Do you have a use case in mind? My goal here was to have a well-defined interface for metrics (`reset`, `update`, `compute`) that doesn't require that developers of new metrics concern themselves with events at all. Getting rid of `started` or `iteration_completed` would mean that event concepts would leak into the metric implementations. What do you think?

Comment (alykhantejani) [reply ignite/metrics/metric.py]: I agree the defined interface of `reset`, `update` and `compute` is really nice and clean. 

Is it the idea that metrics will always be updated every iteration, and only computed at the end of your engine run as I imagine that metrics will also be used in training (not just evaluation) to monitor. Also, in this case when/how should `reset()` be called? 

I guess the notion of "attaching" the metric to an engine already ties it to an event model, but by adding in the flexibility it allows them to be used in training (i.e. reset every epoch/compute every epoch) and the evaluator. By defining the defaults to something sensible means for the common use case of an evaluator it's just an out-of-the-box `metric.attach(evaluator)`?

It's not super clear with the current API how to use it with a Trainer, or if this is a use case?

File: ignite/metrics/metric.py (hunk 2 above)
Comment (alykhantejani) [reply ignite/metrics/metric.py]: maybe option 1? then it's consistent with how you pass it in and how its represented on the `State`?

File: ignite/metrics/metric.py (hunk 1 above)
Comment (jasonkriss) [reply ignite/metrics/metric.py]: Ahh I see. The issue here is that I have built this out (so far at least) to be purely used with evaluators. So i guess we should first take a step back and discuss that...

Should these be just used with evaluators or should they work with trainers as well?

My personal preference is for metrics to be solely an evaluation concern. I think using them with trainers will cause quite a bit of confusion. We would need to modify `create_supervised_trainer` to take a set of metrics and change the return value of the supervised `process_fn`. I think the division of Trainer and Evaluator starts to get a lil confusing if metrics are used on both.

The general (supervised) workflow I've had in mind, is you can log the loss returned by the trainer's `process_fn` to gauge convergence every iteration. But evaluation of metrics happens against a static target (no changing weights during the computation) every so often. The evaluation can certainly be done against the training set as well as an validation set as in the example of above. 

What do you think?

File: ignite/metrics/metric.py (hunk 2 above)
Comment (jasonkriss) [reply ignite/metrics/metric.py]: 👍 I think that sounds great. I'll make the change.

File: ignite/evaluator.py [hunk 3]
```diff
@@ -12,24 +12,25 @@ def add_event_handler(self, event_name, handler, *args, **kwargs):
...
 
 
-def create_supervised_evaluator(model, cuda=False):
+def create_supervised_evaluator(model, metrics={}, cuda=False):
     """
     Factory function for creating an evaluator for supervised models
 
     Args:
         model (torch.nn.Module): the model to train
+        metrics (list of Metrics): a list of metrics to compute
```
Comment (alykhantejani) [inline ignite/evaluator.py]: This should now be dict

File: ignite/metrics/metric.py (hunk 1 above)
Comment (alykhantejani) [reply ignite/metrics/metric.py]: Ah I see what you mean, yes! I really like this approach! If the user wants to run metrics on the training set they can define an evaluator and run it on the training set. 

Perfect, this looks pretty good to me. Thanks @jasonkriss!

File: ignite/evaluator.py (hunk 3 above)
Comment (jasonkriss) [reply ignite/evaluator.py]: good catch

**Comment by vfdev-5** [comment]
Sorry I missed previous discussions on the subject, I wonder whether we can have a wrapping on a custom loss function as a metric for the evaluator ?
In the example [mnist_with_tensorboardx.py](https://github.com/pytorch/ignite/blob/master/examples/mnist_with_tensorboardx.py) we create a trainer with a loss function nn.NLLLoss() and an evaluator that computes metrics:
```python
trainer = create_supervised_trainer(model, optimizer, nn.NLLLoss(), cuda=cuda)
evaluator = create_supervised_evaluator(model,
                                            metrics={'accuracy': CategoricalAccuracy(),
                                                     'nll': NegativeLogLikelihood()}, cuda=cuda)
```
but imagine a situation when the loss function is something custom and we would like also to compute it on the validation.

**Comment by jasonkriss** [comment]
@vfdev-5 I think that's a great idea and something I was considering myself. I'll open a PR right now for y'all to review.

**Comment by jasonkriss** [comment]
@vfdev-5 #112
